### PR TITLE
[SofaMiscTopology] Add a component TopologyBoundingTrasher to remove mesh going outside from scene bounding box

### PR DESCRIPTION
**Pull request by epernod**
Given a link to the topology, link to the corresponding mechanicalObject and a boundingBox. The component will remove all elements (topology and dofs) going outside from this boundingbox.

**-> Any better name suggestion is welcomed.
-> if this component is not useful in SOFA, abort the PR. otherwise I will add more comments in the header.**

Example of use:
```
        <MechanicalObject src="@../grid_1" name="Volume" />
        <TetrahedronSetTopologyContainer name="Tetra_con" src="@../Beam_generator_01/Container"/>
        <TetrahedronSetTopologyModifier name="Modifier" />
        <TetrahedronSetGeometryAlgorithms name="GeomAlgo" template="Vec3d" drawTetrahedra="1" drawScaleTetrahedra="0.8"/>
     
        <TopologyBoundingTrasher box="-40 -40 -50 50 50 50" topology="@Tetra_con" drawBox="1" position="@Volume.position"/>
```

Video of the result of a tissue falling in 3D:
**Right now in SOFA:**
ezgif com-gif-maker(1)

**Using the trasher!**
ezgif com-gif-maker

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build][force-full-build]

**Comment by fredroy** [comment]
Could you add an example scene file in this PR (maybe the one from the animation 😏) ?

**Comment by epernod** [comment]
> 
> 
> Could you add an example scene file in this PR (maybe the one from the animation 😏) ?

Done 😉

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    void reinit() override;
+
+    void handleEvent(sofa::core::objectmodel::Event* event) override;
+
+    void draw(const core::visual::VisualParams* vparams) override;
+
+public:
+    Data<VecCoord> d_positions;
+    Data<Vec6>  d_borders;
+    Data<bool>  m_drawBox; ///< draw bounding box
```
Comment (fredroy) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.h]:
```suggestion
    Data<bool>  d_drawBox; ///< draw bounding box
```

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+protected:
+    core::topology::BaseMeshTopology* m_topology;
+    sofa::core::topology::TopologyElementType m_topologyType;
+
+    sofa::component::topology::EdgeSetTopologyModifier* edgeModifier;
+    sofa::component::topology::TriangleSetTopologyModifier* triangleModifier;
+    sofa::component::topology::QuadSetTopologyModifier* quadModifier;
+    sofa::component::topology::TetrahedronSetTopologyModifier* tetraModifier;
+    sofa::component::topology::HexahedronSetTopologyModifier* hexaModifier;
```
Comment (fredroy) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.h]:
```suggestion
    core::topology::BaseMeshTopology::SPtr m_topology;
    sofa::core::topology::TopologyElementType m_topologyType;

    sofa::component::topology::EdgeSetTopologyModifier::SPtr edgeModifier;
    sofa::component::topology::TriangleSetTopologyModifier::SPtr triangleModifier;
    sofa::component::topology::QuadSetTopologyModifier::SPtr quadModifier;
    sofa::component::topology::TetrahedronSetTopologyModifier::SPtr tetraModifier;
    sofa::component::topology::HexahedronSetTopologyModifier::SPtr hexaModifier;
```

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.inl [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+namespace sofa::component::misc
+{
+
+using namespace sofa::core::topology;
+
+template <class DataTypes>
+TopologyBoundingTrasher<DataTypes>::TopologyBoundingTrasher()
+    : d_positions(initData(&d_positions, "position", "position coordinates of the topology object to interact with."))
+    , d_borders(initData(&d_borders, Vec6(-1000, -1000, -1000, 1000, 1000, 1000), "box", "List of boxes defined by xmin,ymin,zmin, xmax,ymax,zmax"))
+    , m_drawBox(initData(&m_drawBox, false, "drawBox", "Draw Boxes. (default = false)"))
```
Comment (fredroy) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.inl]:
```suggestion
    , d_drawBox(initData(&d_drawBox, false, "drawBox", "Draw Boxes. (default = false)"))
```

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.inl [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+    {
+        filterElementsToRemove();
+    }
+}
+
+
+template <class DataTypes>
+void TopologyBoundingTrasher<DataTypes>::draw(const core::visual::VisualParams* vparams)
+{
+    if (m_drawBox.getValue())
```
Comment (fredroy) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.inl]:
```suggestion
    if (d_drawBox.getValue())
```

**Review by fredroy — COMMENTED** [review]
- using SPtr instead of raw pointers
- guideline for the Data `drawBox`

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.h [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+protected:
+    core::topology::BaseMeshTopology::SPtr m_topology;
+    sofa::core::topology::TopologyElementType m_topologyType;
+
+    sofa::component::topology::EdgeSetTopologyModifier::SPtr edgeModifier;
+    sofa::component::topology::TriangleSetTopologyModifier::SPtr triangleModifier;
+    sofa::component::topology::QuadSetTopologyModifier::SPtr quadModifier;
+    sofa::component::topology::TetrahedronSetTopologyModifier::SPtr tetraModifier;
+    sofa::component::topology::HexahedronSetTopologyModifier::SPtr hexaModifier;
```
Comment (fredroy) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyBoundingTrasher.h]:
```suggestion
    sofa::core::sptr<sofa::component::topology::EdgeSetTopologyModifier> edgeModifier;
    sofa::core::sptr<sofa::component::topology::TriangleSetTopologyModifier> triangleModifier;
    sofa::core::sptr<sofa::component::topology::QuadSetTopologyModifier> quadModifier;
    sofa::core::sptr<sofa::component::topology::TetrahedronSetTopologyModifier> tetraModifier;
    sofa::core::sptr<sofa::component::topology::HexahedronSetTopologyModifier> hexaModifier;
```

**Comment by fredroy** [comment]
Ready !

**Pull request merged**